modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteUncollect.java

```diff
@@ -0,0 +1,121 @@
+ */
+public class IgniteUncollect extends Uncollect implements IgniteRel {
+    /**
+     * Creates an Uncollect relational operator.
+     */
+    public IgniteUncollect(RelOptCluster cluster, RelTraitSet traitSet, RelNode child, boolean withOrdinality) {
+        super(cluster, traitSet, child, withOrdinality, Collections.emptyList());
+    }
+
+    /** */
+    public IgniteUncollect(RelInput input) {
+        super(changeTraits(input, IgniteConvention.INSTANCE));
+    }
+
+    /**
+     * Creates an IgniteUncollect.
+     *
+     * <p>Each field of the input relational expression must be an array or multiset.
+     *
+     * @param traitSet Trait set
+     * @param input    Input relational expression
+     * @param withOrdinality Whether output should contain an ORDINALITY column
+     */
+    public static IgniteUncollect create(RelTraitSet traitSet, RelNode input, boolean withOrdinality) {
+        RelOptCluster cluster = input.getCluster();
+
+        return new IgniteUncollect(cluster, traitSet, input, withOrdinality);
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelNode copy(RelTraitSet traitSet, RelNode input) {
+        return new IgniteUncollect(getCluster(), traitSet, input, withOrdinality);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T> T accept(IgniteRelVisitor<T> visitor) {
+        return visitor.visit(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteRel clone(RelOptCluster cluster, List<IgniteRel> inputs) {
+        return new IgniteUncollect(cluster, getTraitSet(), sole(inputs), withOrdinality);
+    }
+
+    /** {@inheritDoc} */
+    @Override public double estimateRowCount(RelMetadataQuery mq) {
+        int fieldCnt = getInput().getRowType().getFieldCount();
+
+        // Assume every field contains collection of 4 elements (to simplify calculation).
+        long rowsMultiplier = fieldCnt > 15 ? Integer.MAX_VALUE : (1L << (fieldCnt * 2));
+
+        return rowsMultiplier * mq.getRowCount(getInput());
+    }
+
+    /** {@inheritDoc} */
+    @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(RelTraitSet required) {
+        if (required.getConvention() != IgniteConvention.INSTANCE)
+            return null;
+
+        if (TraitUtils.collation(required) != RelCollations.EMPTY)
+            return null;
+
+        if (TraitUtils.distribution(required).getType() == RelDistribution.Type.HASH_DISTRIBUTED)
+            return null;
+
+        return Pair.of(required, ImmutableList.of(required));
+    }
+
+    /** {@inheritDoc} */
+    @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(RelTraitSet childTraits, int childId) {
+        assert childId == 0;
+
+        if (childTraits.getConvention() != IgniteConvention.INSTANCE)
+            return null;
+
+        RelTraitSet traits = childTraits.replace(RelCollations.EMPTY);
+
+        if (TraitUtils.distribution(traits).getType() == RelDistribution.Type.HASH_DISTRIBUTED)
```

Comment:
I commented these lines and none of the new tests failed.

Reply:
See related comment about collation. For deriveTraits it will be wierd to have hash distribution on collection field, but in case someone is used it we should process it correctly. For passThroughTraits such distribution can be requested by downstream node, and we can't provide it without exchange (or trim exchange in case of brodcast distribution)